enhancements/sig-policy/74-order-policy-execution/README.md

```diff
@@ -0,0 +1,227 @@
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [website](https://github.com/open-cluster-management-io/open-cluster-management-io.github.io/)
+
+## Summary
+
+Multiple policy templates are often wrapped in one policy, but all this does is create
+all of the policy templates at the same time. In any use case where a policy template 
+depends on another policy being processed before it, the only way to guarantee the policy
+templates are processed in the correct order is to split them into separate policies and
+create them in the desired order. We propose allowing users to specify conditions that
+must be met in order to evaluate a policy template, which would allow the governance
+framework to automatically handle cases where policy templates need to be processed in
+order.
+
+## Motivation
+
+Currently, users have to split up policy templates that depend on others into multiple
+policies and manually create them in order. This is more resource intensive as users
+can have to create duplicate policies and more time consuming if users have to manually
+manage policy processing order. If a user elects to group the templates in one policy,
+there is no guarantee the templates will be processed in order, which could leave
+the policy in a bad state if a required resource has not yet been processed. For example,
+when using a policy to install an operator, if some objects are created before others it
+can cause the operator to not be installed properly. Allowing users to specify policy
+dependencies will alleviate these issues.
+
+### Goals
+
+1. Policy template dependencies: allow users to automatically process policy templates
+in a desired order and guarantee that this order will be stable
+2. Opt-in: any extra fields that are used to specify policy ordering should be optional,
+and policies that do not contain these fields should function as they did before in order
+to avoid issues with older policies
+
+### Non-Goals
+
+1. Status message matching: we plan to only allow users to conditionally process policy
+templates based on compliance state, so users will not be able to specify that they want a
+policy to be processed if another policy has a specific status message
+
+## Proposal
+
+We would add an optional new `dependencies` field to the spec on the configuration 
+policy level as well as an `extraDependencies` field on the template level to
+allow users to specify conditions for policy processing on both the policy and
+template level. When these fields are set and their conditions are not met, 
+the policy templates will not be applied on the managed cluster and the affected
+policy will enter a "pending" state. If this happens, any policy sets that contain
+the pending policy will also be marked as pending. When the conditions are met or
```

Comment:
I had an extra thought around when to put the whole policy into a Pending state: since it could be that only one policy template is Pending, but others are Compliant and NonCompliant, which state gets the highest priority? As-is, it sounds like the Pending status would override any others, but I wonder if NonCompliant should be the most important...

Reply:
thinking about it more, you are probably right - I think users would want to know if any policies are noncompliant at a glance even if parts of them are still pending. but I think we should prioritize pending over compliant, since we don't want to report compliant if some templates have not been processed yet

Reply:
I agree with what @willkutler said.

Reply:
One more extra thought (sorry): if I make a policy that I want to "activate" when another thing is NonCompliant (or something), does it make sense to treat that as "Pending" in the same way as the others?

I imagined a policy like that would be used to do some additional automatic remediation on something, in which case it's "normal" state would probably be non-active. So should it be preventing the parent Policy from becoming Compliant?

What if (optionally) a pending template could be ignored in the Policy's compliance? Like:
```
apiVersion: policy.open-cluster-management.io/v1
kind: Policy
metadata:
  name: my-complicated-policy
spec:
  policy-templates:
    - extraDependencies:
        - kind: ConfigurationPolicy
          name: bar
          compliance: NonCompliant
      ignorePending: true // <===== New <=====
      objectDefinition:
        apiVersion: policy.open-cluster-management.io/v1
        kind: ConfigurationPolicy
        metadata:
          name: fix-bar-in-foo
        spec:
          remediationAction: enforce
          ...
```

Reply:
yeah, I like that idea @JustinKuli . I definitely think there are some scenarios where we basically want to ignore a policy template unless something bad happens, and in those cases we wouldn't want it messing up the status.